src/runtime/pipeline/pipeline_struct.h

```diff
@@ -130,41 +209,319 @@ struct OutputMap {
+   * \param reader Json reader.
+   */
+  void Load(dmlc::JSONReader* reader) {
+    reader->BeginArray();
+    while (reader->NextArrayItem()) {
+      std::string key;
+      reader->BeginObject();
+      int mod_idx = -1;
+      ConfigOutputBindings output;
+      std::string dev;
+      while (reader->NextObjectItem(&key)) {
+        if (key == "mod_idx") {
+          reader->Read(&mod_idx);
+        } else if (key == "dev") {
+          reader->Read(&dev);
+        } else if (key == "output") {
+          reader->Read(&output);
+        } else {
+          LOG(FATAL) << "do not support key " << key;
+        }
+      }
+      ICHECK(mod_idx >= 0) << "Invalid mod_idx value " << mod_idx;
+      // Check if the output is successfully read.
+      ICHECK(!output.Empty()) << "Invalid output binding result.";
+      // Build the mapping of mod_idx and "ConfigOutputBindings".
+      config_[mod_idx] = output;
+    }
+    // Call this function after "config" loading finished.
+    ParseConfiguration(config_);
+  }
+};
+/*
+ *\brief Runtime of backend.
+ */
+class BackendRuntime {
+ private:
+  /*\brief The index of runtime indicate the position in the pipeline.*/
+  int runtime_idx_;
+  /*\brief The Runtime module of a backedn graph executor.*/
+  Module module_;
+  /*!
+   *\brief To transfer data between two different backends, we need a local
+   * tensor variable as a medium. This variable is a mapping of input data and local
+   * data.
+   */
+  std::unordered_map<DLTensor*, DLTensor*> input_tensor_local_copy_;
+  /*!\brief The packed functions.*/
+  tvm::runtime::PackedFunc run_;
+  tvm::runtime::PackedFunc set_input_;
+  tvm::runtime::PackedFunc get_input_;
+  tvm::runtime::PackedFunc get_output_;
+  tvm::runtime::PackedFunc get_num_output_;
+  tvm::runtime::PackedFunc get_num_inputs_;
+  tvm::runtime::PackedFunc get_input_index_;
+  /*!
+   * \brief The new DLTensor have same shape, data type with a existing DLTensor and
+   *  the new DLTensor device type is CPU. This function is used to do cross device data
+   *  copy.
+   */
+  inline DLTensor* CopyDLTensorToCPU(const DLTensor* from) {
+    DLTensor* ret = NULL;
+    TVMArrayAlloc(from->shape, from->ndim, from->dtype.code, from->dtype.bits, from->dtype.lanes,
+                  kDLCPU, 0, &ret);
+    return ret;
+  }
+  /*!\brief The new NDArray have same shape, data type with an existing DLTensor.*/
+  NDArray CreateNDArrayFromDLTensor(const DLTensor* from) {
+    std::vector<int64_t> shape;
+    for (int i = 0; i < from->ndim; i++) {
+      shape.push_back(from->shape[i]);
+    }
+    auto ndarray = NDArray::Empty(shape, from->dtype, from->device);
+    ndarray.CreateView(shape, from->dtype);
+    return ndarray;
+  }
+  /*
+   *\brief Copy data from a DLTensor to another DLTensor.
+   */
+  void CopyFromTo(DLTensor* from, DLTensor* to) {
+    // If the source device and target device is not same, we use a local DLTensor
+    // as a medium to do the cross device copy work.
```

Comment:
```suggestion
    // If the source device and target device are not the same, we use a temporary DLTensor on CPU
    // as the bridge.
```
After thinking twice, I feel this may still introduce unnecessary overhead. Isn't it possible for non-CPU device APIs to support `from`? Should we first try `TVMArrayCopyFromTo(from, to, nullptr)`, and then go through CPU if failed?

Reply:
Current existing TVMArrayCopyFromTo  logic  supported `from` device, once two device both are non-cpu, the said copy function will calling `from` DeviceAPI function to copy data from `from` to `to`, in such case reading data from `from` device is not problem, but write data to `to` device may cause crash issue. the reason is  the `from` device may have no capability to access the ```data``` of  ```DLTensor of to```, for example when `to` device  is VTA, the memory address in DLTensor actually is the address of data structure VTABuffer pointer instead of data memory address,  a try of `TVMArrayCopyFromTo`  to directly copy data into the  `to` DLTensor  will cause a crash  by  memory access violation. 

if we revise the logic to let TVMArrayCopyFromTo to support `to` device, then read data from  `from` device would become a new problem, hence here we use a bridge NDArray as a solution.